### DISCUSSION TITLE
Plugin development: bubble up validation error

### DISCUSSION THREAD

**Discussion by thegcat** [pretix/pretix#3418] [Q&A]
The [django docs](https://docs.djangoproject.com/en/3.2/ref/forms/validation/#cleaning-a-specific-field-attribute) suggest that it should be possible to pass a `ValidationError` to a `Form` from a `clean_attribute` method.

We are unfortunately unable to make this work, see [our plugin](https://gitlab.fachschaften.org/kif/pretix-matrix-inviter/-/blob/main/pretix_matrix_inviter/forms.py#L69). Is this a limitation in Pretix or are we holding it wrong?

**Comment by raphaelm** [comment]
I'm not aware of such a limitation in pretix and with a 5-minute glance, I don't see an error with your code

**Comment by wiffbi** [comment]
As `urllib3.util.parse_url` is very forgiving when parsing URLs (see e.g. incomplete URLs [here](https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html#urllib3.util.parse_url)), maybe everything works as it should, but you expect something different? Do you e.g. want it to fail, if no host is given?

**Reply by thegcat** [reply]
Yes, I know `urrllib3.util.parse_url` is very forgiving, but I tried for example `matrix...`, which funnily enough threw `LocationParseError` somewhere later in the code that used `requests` with the hostname, but still didn't see an error message on the form.

Anyway, I'll try to debug this a bit more, and if I'm satisfied the `raise` raises correctly I might come back with an error report 😅

**Reply by wiffbi** [reply]
Interesting, when I run the following, it happily allows `"matrix..."`

```
from urllib3.util import parse_url
>>> parse_url('matrix...')
Url(scheme=None, auth=None, host='matrix...', port=None, path=None, query=None, fragment=None)
```

Let me/us know what you find. :)

**Reply by thegcat** [reply]
Ah, the error thrown on trying the request was `LocationParseError` too, though from somewhere else in `urllib3`, see below.

Is there anything else you could suggest to test the well-formedness of a hostname, with optionally scrubbing the protocol and whatever other things from the input? I could just query the matrix server's version or anything else, but that would open the validation to other errors like flaky connections, hiccuping matrix servers and so on.

```
  File "/var/pretix/venv/lib/python3.10/site-packages/requests/api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
  File "/var/pretix/venv/lib/python3.10/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/var/pretix/venv/lib/python3.10/site-packages/requests/sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
  File "/var/pretix/venv/lib/python3.10/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "/var/pretix/venv/lib/python3.10/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/var/pretix/venv/lib/python3.10/site-packages/urllib3/connectionpool.py", line 790, in urlopen
    response = self._make_request(
  File "/var/pretix/venv/lib/python3.10/site-packages/urllib3/connectionpool.py", line 467, in _make_request
    self._validate_conn(conn)
  File "/var/pretix/venv/lib/python3.10/site-packages/urllib3/connectionpool.py", line 1092, in _validate_conn
    conn.connect()
  File "/var/pretix/venv/lib/python3.10/site-packages/urllib3/connection.py", line 604, in connect
    self.sock = sock = self._new_conn()
  File "/var/pretix/venv/lib/python3.10/site-packages/urllib3/connection.py", line 200, in _new_conn
    sock = connection.create_connection(
  File "/var/pretix/venv/lib/python3.10/site-packages/urllib3/util/connection.py", line 58, in create_connection
    raise LocationParseError(f"'{host}', label empty or too long") from None
urllib3.exceptions.LocationParseError: Failed to parse: 'matrix..', label empty or too long
```

**Reply by wiffbi** [reply]
You could try and validate the hostname in your `clean_matrix_inviter_matrix_server` the same way the error is thrown in `create_connection` in urllib3. E.g. somethin like this (sorry, untested)

```
    def clean_matrix_inviter_matrix_server(self):
        url = self.cleaned_data["matrix_inviter_matrix_server"]
        try:
            url = parse_url(url).host
            url.encode("idna") # throws UnicodeError if hostname is not valid
        except (LocationParseError, UnicodeError):
            raise ValidationError("Please enter a valid hostname")
        return url

```

**Reply by raphaelm** [reply]
I'd suggest using django's URLValidator kr URLField and *then* extract the hostname with urllib if you need to
